base/cloudant_base.go

```diff
@@ -195,6 +195,15 @@ func (c *BaseService) SetServiceURL(url string) error {
 // the retryable client; otherwise "client" will be stored
 // directly on "service".
 func (c *BaseService) SetHTTPClient(client *http.Client) {
+	if client.Transport == nil {
+		client.Transport = http.DefaultTransport
+	}
+
+	// wrap client's transport into ErrorResponse unless it was already wrapped
+	if _, ok := client.Transport.(*ErrorResponse); !ok {
```

Comment:
Is this check robust to our `ErrorResponse` `Transport` being wrapped inside other potentially user-provided `Transport`s?

Reply:
Not really, because if user's wrapping our `ErrorResponse` in their own `Transport` implementation detecting depends on implementation details which we don't know. I tried to introduce `Middleware` interface where we'd only accept `Transports` with wrapping done our way, but that reverses problem, because reflection wouldn't show if given middleware is `ErrorResponse` or if it is somewhere in the chain.

So this check here is best we can do without digging ourselves too deep - it protects us from double-wrapping when user takes our `Transport`, tweaks something on it, say, timeout, and puts it back. And it is wrapping completely separate user specified `Transports` in our `ErrorResponse`.

If it is user-provided `Transport` which is also a middleware that wraps our `ErrorResponse` I hope that'd indicate they know what they are doing and ready to pay price of double detection on errors, because it wouldn't go beyond that, error aug is idempotent, after all.

Reply:
Yeah OK that's fair, thanks for the explanation.